Revisão Migrations e afins

git-jr opened the pull request.

git-jr committed:
Adicionado dialog "ListaUsuários"

git-jr committed:
Adição de Migrations manuais, automática e específicas

git-jr committed:
ajuste no layout da dialog que lista usuarios

git-jr committed:
filtragem por contas adicionado
- Só está funcionando, requer refatorações

git-jr committed:
tela de busca de contatos adicionada

git-jr committed:
Mudança de abordagem com buscas vazias e o operador LIKE

git-jr committed:
Adição de telas: detalhes e edição de usuário

git-jr committed:
Exclusão e edição com Chave Estrangeira

git-jr committed:
Ajustes

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/navigation/ListaUsuariosNavigation.kt:
```diff
@@ -0,0 +1,108 @@
+import br.com.alura.helloapp.util.USUARIO_ATUAL
+import kotlinx.coroutines.launch
+
+fun NavGraphBuilder.listaUsuarios(
```
alterar para grafo pois tem mais de 1 destino.

git-jr replied:
Alterado

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/navigation/ListaUsuariosNavigation.kt:
```diff
@@ -0,0 +1,108 @@
+                }
+            },
+            onClickApagar = {
+                coroutineScope.launch {
```
tentar implementar o viewmodel global para a tela não ser responsável em redirecionar para login ao 'deslogar'

git-jr replied:
feito!

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/HelloAppNavHost.kt:
```diff
@@ -25,6 +23,8 @@ fun HelloAppNavHost(
         homeGraph(navController)
```
talvez elevar os eventos de navegação para o navhost https://github.com/alura-cursos/jetpack-compose-navigation-deep-links-e-testes/blob/projeto-inicial/app/src/main/java/br/com/alura/panucci/navigation/PanucciNavHost.kt

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/HelloAppNavHost.kt:
```diff
@@ -52,4 +52,12 @@ fun NavHostController.navegaParaFormularioContato(idContato: Long = 0L) {
     navegaDireto(DestinosHelloApp.LoginGraph.rota)
+}
+
+fun NavHostController.navegaParaDialgoUsuarios(idUsuarioAtual: String = "") {
```
dado que a navegação com argumento não é opcional, os parametros das funções tbm não devem ser

git-jr replied:
Corrigido, tanto nesse arquivo de NavHost quanto no "DestinosHelloApp" removendo `defaultValue` de `navArgument`

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/BuscaContatosTela.kt:
```diff
@@ -0,0 +1,114 @@
+@Composable
+fun BuscaContatosTela(
+    state: BuscaContatosUiState,
+    modifier: Modifier = Modifier,
```
adicionar o modifier no parent maior, o Scaffold

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/BuscaContatosTela.kt:
```diff
@@ -0,0 +1,114 @@
+                )
+            }
+
+            BasicTextField(
```
em curso, comentar sobre a personalização a partir de basic.

git-jr replied:
anotado ✔

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/BuscaContatosTela.kt:
```diff
@@ -0,0 +1,114 @@
+
+@Preview
+@Composable
+fun AppBarBuscaContatosPreview() {
```
preview gambeta

git-jr replied:
Corrigido kk

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/BuscaContatosUiState.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class BuscaContatosUiState(
+    val contatos: List<Contato> = emptyList(),
+    val usuarioAtual: String = "",
```
reforçando o usuário atual ser nullable

git-jr replied:
Arrumei esse, mas vou revisar o código depois para procurar casos semelhantes

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/ListaContatosUiState.kt:
```diff
@@ -4,5 +4,6 @@ import br.com.alura.helloapp.data.Contato
 data class ListaContatosUiState(
     val contatos: List<Contato> = emptyList(),
-    val logado: Boolean = true
+    val logado: Boolean = true,
```
parece contra intuitivo o padrão ser sempre logado, pois o logado é uma situação condicional, sendo assim, não seria melhor começar com não logado e provar o contrário?

git-jr replied:
Curioso, sinceramente nem me lembro porquê setei esse valor como true originalmente, em outros pontos do código eu  uso false por padrão para propriedade `logado`. Inclusive, dado que os comportamentos de Login serão feitos por uma navegação aparte, essa property não será mais necessária aqui.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/login/FormularioLoginViewModel.kt:
```diff
@@ -39,16 +42,36 @@ class FormularioLoginViewModel @Inject constructor(
                         senha = it
                     )
                 },
+                voltarParaLoginMudou = {
```
provavelmente isso tem relação com o comportamento de deslogar, talvez seja melhor deixar isso apenas no viewmodel focado em autenticação, usuário etc.

git-jr replied:
Esse comportamento em específico eu não sei exatamente como linkar como o viewModel global.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/login/FormularioLoginViewModel.kt:
```diff
@@ -39,16 +42,36 @@ class FormularioLoginViewModel @Inject constructor(
+                    senha = senha
+                )
+            )
+            if (insere == -1L) { // Garaças a @Insert(onConflict = IGNORE), -1 é "já existe"
```
mudar para try catch e lidar diretamente com exception

git-jr replied:
alterado

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/login/LoginViewModel.kt:
```diff
@@ -46,20 +48,23 @@ class LoginViewModel @Inject constructor(
-            if (usuario == _uiState.value.usuario &&
-                senha == _uiState.value.senha
-            ) {
+        if (usuarioBuscado != null) {
```
se preferir, pode usar ?.let por aqui

git-jr replied:
Blz, aqui eu optei pelo _if  != null,_ para usar o _else_ depois

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+    suspend fun atualizar() {
```
matenha apenas um padrão para verbos ou imperativo ou infinitivo.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,113 @@
+            mostraMensagemExclusaoMudou(false)
+        }
+
+        verificaRotaDeVolta()
```
provavelmente é uma outra lógica que vai ser feita apenas no vm que cuida de user

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/ListaUsuariosTela.kt:
```diff
@@ -0,0 +1,256 @@
+import br.com.alura.helloapp.ui.theme.HelloAppTheme
+
+@Composable
+fun CaixaDialogoContasUsuario(
```
Lembrei que ainda vai ter algumas mudanças

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/ListaUsuariosViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+        }
+
+        usuarioAtual?.let {
+            usuarioDao.buscaTodos().collect { usuariosBuscados ->
```
lembrei que é aquele caso que dá pra resolver em query

alexfelipe reviewed (changes requested):
Fiz a revisão e deixei os pontos de mudança, de uma forma geral, acho que vale se atentar aos detalhes para ajustar em todos os pontos que aparecer. Dado que pretende usar o Type Safety, é melhor aplicar em tudo, seja em cada destino, como também, nos métodos de navegação. Outro ponto importante é garantir que apenas o viewmodel de usuário vai ser o responsável em lidar com autenticação e direcionamento da navegação, sendo assim, o restante das telas vai apenas direcionar que não tem mais usuário a partir do datastore no máximo

git-jr committed:
pequenos ajustes

git-jr committed:
Adição de ViewModel Global para gerencial sessao

git-jr committed:
Elevação dos eventos de navegação para o Navhost

git-jr committed:
Correção de rotas e ajustes de nomenclatura

git-jr commented on app/src/main/java/br/com/alura/helloapp/ui/HelloAppNavHost.kt:
```diff
@@ -1,31 +1,78 @@
+            navController.navegaParaDetalhes(idContato)
+        })
     }
+
```
ViewModel Global

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/database/HelloAppDatabase.kt:
```diff
@@ -1,13 +1,58 @@
-@Database(entities = [Contato::class], version = 1)
+@Database(
+    entities = [Contato::class, Usuario::class],
+    version = 13,
```
só pra confirmar, vai manter todas as migrations mesmo ou vai reduzir? Se forem 13, provavelmente vai ficar muito complexo para o curso

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/navigation/BuscaGraphNavigation.kt:
```diff
@@ -0,0 +1,33 @@
+import br.com.alura.helloapp.ui.navegaParaDialgoUsuarios
+
+fun NavGraphBuilder.buscaContatosGraph(
+    onClickVoltar: () -> Unit,
```
nessa parte o evento não fica atrelado à ação da UI, ou seja, apenas indicar que vai voltar ou vai navegar para algum ponto específico é o suficiente.

git-jr replied:
Ok, faz sentido! Nesse caso em especifico o evento de voltar depende do click no icone de voltar `<-` da AppBar, mas em outras telas vou ficar atento a isso.

git-jr replied:
Alterei outros pontos do código seguindo esse raciocínio.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/navigation/DetalhesContatoNavigation.kt:
```diff
@@ -18,7 +18,8 @@ import br.com.alura.helloapp.util.ID_CONTATO
 
 fun NavGraphBuilder.detalhesContatoGraph(
-    navController: NavHostController
+    onClickVoltar: () -> Unit,
```
marcando mais uma vez o vinculo da ação de ui com o evento para fazer a navegação, se atente aos outros pontos também

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/HelloAppNavHost.kt:
```diff
@@ -1,31 +1,78 @@
-        homeGraph(navController)
-        formularioContatoGraph(navController)
-        detalhesContatoGraph(navController)
+        splashGraph(onNavegaParaLogin = {
```
tenta formatar para ficar em cada linha

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/SessaoViewModel.kt:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    private suspend fun monitoraLogin() {
```
manter apenas uma chamada dado que ela apenas invoca outro método

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/SessaoViewModel.kt:
```diff
@@ -0,0 +1,58 @@
+    private val dataStore: DataStore<androidx.datastore.preferences.core.Preferences>
+) : ViewModel() {
+
+    private val _uiState = MutableStateFlow(SessaoState())
```
nesse caso não seria um ui state

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/home/ListaContatosUiState.kt:
```diff
@@ -4,5 +4,5 @@ import br.com.alura.helloapp.data.Contato
 data class ListaContatosUiState(
     val contatos: List<Contato> = emptyList(),
-    val logado: Boolean = true
+    val usuarioAtual: String = ""
```
nullable

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/login/FormularioLoginViewModel.kt:
```diff
@@ -39,16 +39,34 @@ class FormularioLoginViewModel @Inject constructor(
+                        voltarParaLogin = it
+                    )
+                },
+                onErro = {
```
alterar para modificar diretamente a property, pois o evento de mudança não ocorre na tela. Tentar explicar isso no video de mudanças/melhorias do projeto

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/login/LoginViewModel.kt:
```diff
@@ -45,21 +48,23 @@ class LoginViewModel @Inject constructor(
                 _uiState.value.onErro(true)
             }
+        } else {
+            _uiState.value.onErro(true)
```
Dado que existem duas situações diferentes para a falha, talvez faça sentido especificar o problema que aconteceu e indicar uma mensagem mais precisa na tela

git-jr replied:
Optei por manter a mesma mensagem de erro, porém, modifiquei o código para simplificar a estrutura de decisão.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+    private val dataStore: DataStore<Preferences>
+) : ViewModel() {
+
+    private val nomeUsuario = savedStateHandle.get<String>(br.com.alura.helloapp.util.USUARIO_ATUAL)
```
import static para simplificar

alexfelipe replied:
Dado que existe um com o mesmo nome, tente usar um nome diferente

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+            usuario.collect { usuarioBuscado ->
+                usuarioBuscado?.let {
+                    _uiState.value = _uiState.value.copy(
+                        nomeUsuario = it.nomeDeUsuario, nome = it.nome, senha = it.senha.toString()
```
acho que é bom evitar buscar a senha e mostrá-la para o usuário final.

git-jr replied:
Ok, nova abordagem aplicada: a senha ainda é buscada para usar na anotação `@Update` do Room (para fins didáticos sobre mostrar como o update funciona), mas ela não exibida ao usuário final.

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+                Usuario(
+                    nomeDeUsuario = nomeUsuario,
+                    nome = nome,
+                    senha = senha,
```
caso mudar, tomar cuidado com a atualização da senha, mudar apenas se a pessoa clicar em um botão de mudar a senha ou inserir uma nova

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/FormularioUsuarioViewModel.kt:
```diff
@@ -0,0 +1,106 @@
+        with(_uiState.value) {
+            usuarioDao.apaga(
+                Usuario(
+                    nomeDeUsuario = nomeUsuario,
```
investigar se precisa de todos os campos para remover, acho que só a primary key é o suficiente

git-jr replied:
Realmente só a primary key já basta

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/ListaUsuariosViewModel.kt:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    private suspend fun carregaDados() {
+        val usuario = usuarioAtual?.let { usuarioDao.buscaPorNomeDeUsuario(it).first() }
```
tente colocar toda a lógica dentro desse let

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/ui/userDialog/ListaUsuariosViewModel.kt:
```diff
@@ -0,0 +1,62 @@
+        usuarioAtual?.let {
+            usuarioDao.buscaTodos()
+                .filterNotNull().collect { usuariosBuscados ->
+                    usuariosBuscados.remove(usuario)
```
tente aplicar o filtro na query para devolver todos os users menos esse recebido via argumento

git-jr replied:
feito

alexfelipe reviewed (changes requested):
mais alguns detalhes pra revisar

git-jr committed:
Ajustes de navegação e nomenclatura

git-jr committed:
Inspeção de Código com AS e correção de pequeno bug
- Inspeção do AS para retirar referências e arquivos não utilizados
- O bug fazia com que o formulário de edição de Contato não buscasse corretamente as informações pelo id do usuário logado

alexfelipe commented on app/src/main/java/br/com/alura/helloapp/navigation/BuscaGraphNavigation.kt:
```diff
@@ -0,0 +1,28 @@
+import br.com.alura.helloapp.ui.home.BuscaContatosViewModel
+
+fun NavGraphBuilder.buscaContatosGraph(
+    onClickVolta: () -> Unit,
```
para o host não importa a ação, apenas para onde precisa navegar

alexfelipe reviewed (approved):
parece no esquema
